File: typescript/packages/host/src/cron.ts
import { err, ok, type Result } from "@threads/core/host";

// Five-field cron (minute hour day-of-month month day-of-week) in an IANA time zone (v2
// F10). Wall-clock rules: a nonexistent local time (spring forward) runs at the first valid
// instant after it; an ambiguous one (fall back) runs once, at its first instance.

export type Cron = {
  readonly fields: readonly (ReadonlySet<number> | undefined)[];
};

const RANGES: readonly (readonly [number, number])[] = [
  [0, 59],
  [0, 23],
  [1, 31],
  [1, 12],
  [0, 7],
];

export function parseCron(expr: string): Result<Cron, string> {
  const parts = expr.trim().split(/\s+/);
  if (parts.length !== 5) return err(`cron needs 5 fields: ${expr}`);
  const fields: (ReadonlySet<number> | undefined)[] = [];
  for (const [i, part] of parts.entries()) {
    const [lo, hi] = RANGES[i] ?? [0, 0];
    const set = field(part, lo, hi);
    if (set === null) return err(`bad cron field ${part}`);
    // Sunday is 0 or 7.
    if (i === 4 && set?.has(7) === true) set.add(0);
    fields.push(set);
  }
  return ok({ fields });
}

/** A field's values; undefined for `*` (unrestricted); null when malformed. */
function field(
  part: string,
  lo: number,
  hi: number,
): Set<number> | undefined | null {
  if (part === "*") return undefined;
  const out = new Set<number>();
  for (const item of part.split(",")) {
    const range = rangeOf(item, lo, hi);
    if (range === null) return null;
    for (let v = range.from; v <= range.to; v += range.step) out.add(v);
  }
  return out;
}

/** `*`, `n`, `a-b`, each with an optional `/step`. */
function rangeOf(
  item: string,
  lo: number,
  hi: number,
): {
  readonly from: number;
  readonly to: number;
  readonly step: number;
} | null {
  const m = /^(\*|\d+)(?:-(\d+))?(?:\/(\d+))?$/.exec(item);
  if (m === null) return null;
  const star = m[1] === "*";
  const from = star ? lo : Number(m[1]);
  const open = star || m[3] !== undefined ? hi : from;
  const to = m[2] === undefined ? open : Number(m[2]);
  const step = m[3] === undefined ? 1 : Number(m[3]);
  return from < lo || to > hi || from > to || step < 1
    ? null
    : { from, to, step };
}

type Wall = {
  readonly minute: number;
  readonly hour: number;
  readonly day: number;
  readonly month: number;
  readonly weekday: number;
};

function matches(cron: Cron, w: Wall): boolean {
  const [minute, hour, dom, month, dow] = cron.fields;
  const has = (s: ReadonlySet<number> | undefined, v: number): boolean =>
    s === undefined || s.has(v);
  if (!has(minute, w.minute) || !has(hour, w.hour) || !has(month, w.month))
    return false;
  // Standard cron: with both day fields restricted, either may match.
  if (dom !== undefined && dow !== undefined)
    return dom.has(w.day) || dow.has(w.weekday);
  return has(dom, w.day) && has(dow, w.weekday);
}

/** The zone's wall clock at `ms`, as a UTC-epoch number of the same fields (minute precision). */
function localEpoch(format: Intl.DateTimeFormat, ms: number): number {
  const parts: Record<string, number> = {};
  for (const p of format.formatToParts(ms))
    if (p.type !== "literal") parts[p.type] = Number(p.value);
  return Date.UTC(
    parts["year"] ?? 0,
    (parts["month"] ?? 1) - 1,
    parts["day"] ?? 1,
    parts["hour"] ?? 0,
    parts["minute"] ?? 0,
  );
}

function wall(local: number): Wall {
  const d = new Date(local);
  return {
    minute: d.getUTCMinutes(),
    hour: d.getUTCHours(),
    day: d.getUTCDate(),
    month: d.getUTCMonth() + 1,
    weekday: d.getUTCDay(),
  };
}

const MINUTE = 60_000;

/** Occurrence instants (UTC ms) in (from, to], minute by minute in the zone's wall clock. */
export function occurrences(
  cron: Cron,
  timezone: string,
  from: number,
  to: number,
): readonly number[] {
  const format = new Intl.DateTimeFormat("en-US", {
    timeZone: timezone,
    hourCycle: "h23",
    year: "numeric",
    month: "numeric",
    day: "numeric",
    hour: "numeric",
    minute: "numeric",
  });
  const out: number[] = [];
  const seen = new Set<number>();
  let t = Math.floor(from / MINUTE) * MINUTE + MINUTE;
  let prev = localEpoch(format, t - MINUTE);
  for (; t <= to; t += MINUTE) {
    const local = localEpoch(format, t);
    if (repeated(format, t, local)) seen.add(local);
    if (due(cron, prev, local, seen)) out.push(t);
    prev = local;
  }
  return out;
}

// ponytail: looks back 3 h for the clock turning back; zones whose offset drops by more than
// that in one step would need a longer look-back.
const LOOK_BACK = 3 * 60 * MINUTE;

/** Whether wall time `local` already happened at an earlier instant (a fall-back repeat), even
 *  when that earlier instant is before the scan started. */
function repeated(
  format: Intl.DateTimeFormat,
  t: number,
  local: number,
): boolean {
  const drop =
    localEpoch(format, t - LOOK_BACK) - (t - LOOK_BACK) - (local - t);
  return drop > 0 && localEpoch(format, t - drop) === local;
}

/** This instant fires: its wall minute matches for the first time, or a skipped one did. */
function due(
  cron: Cron,
  prev: number,
  local: number,
  seen: Set<number>,
): boolean {
  let fire = false;
  // Spring forward: wall minutes between prev and local never happened; they run now.
  for (let skipped = prev + MINUTE; skipped < local; skipped += MINUTE)
    if (matches(cron, wall(skipped)) && !seen.has(skipped)) {
      seen.add(skipped);
      fire = true;
    }
  if (matches(cron, wall(local)) && !seen.has(local)) {
    seen.add(local);
    fire = true;
  }
  return fire;
}

File: python/src/threads/host/cron.py
"""Cron expressions (spec/api.json `Schedule.cron`) read in an IANA zone: which UTC minutes an
expression fires at, by the wall-clock rules across daylight-saving changes."""

from dataclasses import dataclass
from datetime import UTC, datetime, timedelta
from typing import Final
from zoneinfo import ZoneInfo, ZoneInfoNotFoundError

from threads.agents.config import ConfigError

MINUTE_MS: Final = 60_000
_RANGES: Final = ((0, 59), (0, 23), (1, 31), (1, 12), (0, 6))
"""minute, hour, day of month, month, day of week (0 is Sunday; 7 is accepted as Sunday)."""


@dataclass(frozen=True, slots=True)
class Cron:
    fields: tuple[frozenset[int], ...]
    dom_any: bool
    dow_any: bool

    def matches(self, at: datetime) -> bool:
        minute, hour, dom, month, dow = self.fields
        weekday = (at.weekday() + 1) % 7
        day = (at.day in dom, weekday in dow)
        # Vixie cron: with both day fields restricted, either one matching is enough.
        day_ok = all(day) if self.dom_any or self.dow_any else any(day)
        return at.minute in minute and at.hour in hour and at.month in month and day_ok


def parse_cron(expression: str) -> Cron:
    """Five fields of numbers, `*`, lists, ranges and steps. Raises ConfigError otherwise."""
    parts = expression.split()
    if len(parts) != len(_RANGES):
        raise ConfigError("invalid_config", f"cron {expression!r} needs five fields")
    fields = tuple(
        _field(p, lo, hi, expression) for p, (lo, hi) in zip(parts, _RANGES, strict=True)
    )
    dow = frozenset(0 if d == 7 else d for d in fields[4])  # noqa: PLR2004 - 7 is Sunday too
    return Cron((*fields[:4], dow), parts[2] == "*", parts[4] == "*")


def _field(text: str, lo: int, hi: int, expression: str) -> frozenset[int]:
    values: set[int] = set()
    top = 7 if hi == 6 else hi  # noqa: PLR2004 - day of week accepts 7
    for item in text.split(","):
        base, _, step_text = item.partition("/")
        try:
            step = int(step_text) if step_text else 1
            if base == "*":
                start, end = lo, hi
            elif "-" in base:
                first, last = base.split("-", 1)
                start, end = int(first), int(last)
            else:
                start = int(base)
                end = hi if step_text else start
        except ValueError:
            raise ConfigError(
                "invalid_config", f"cron {expression!r}: bad field {text!r}"
            ) from None
        if not (lo <= start <= end <= top) or step < 1:
            raise ConfigError("invalid_config", f"cron {expression!r}: {text!r} out of range")
        values.update(range(start, end + 1, step))
    return frozenset(values)


def zone(name: str) -> ZoneInfo:
    try:
        return ZoneInfo(name)
    except (ZoneInfoNotFoundError, ValueError):
        raise ConfigError("invalid_config", f"unknown timezone {name!r}") from None


def is_due(cron: Cron, timezone: str, at: int) -> bool:
    """Whether the instant `at` (UTC ms, on a minute) fires, by the wall-clock rules: a local time
    skipped by a spring-forward runs at the first valid instant after it; a local time repeated by
    a fall-back runs only at its first instance (fold 0)."""
    tz = zone(timezone)
    here = datetime.fromtimestamp(at / 1000, UTC).astimezone(tz)
    before = datetime.fromtimestamp((at - MINUTE_MS) / 1000, UTC).astimezone(tz)
    wall, skipped = here.replace(tzinfo=None, fold=0), before.replace(tzinfo=None, fold=0)
    step = timedelta(minutes=1)
    skipped += step
    while skipped < wall:  # wall minutes that never happened: they run now
        if cron.matches(skipped):
            return True
        skipped += step
    return here.fold == 0 and cron.matches(wall)
